### PR TITLE
Next adaptive exercise button

### PR DESCRIPTION
**Pull request by fogre**
Button to download and open random adaptive exercise.

### CODE REVIEW

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Unnecessary comment.

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import org.openide.util.NbBundle.Messages;
+import org.netbeans.api.project.ui.OpenProjects;
+
+@ActionID(category = "TMC", id = "fi.helsinki.cs.tmc.actions.DownloadAdaptiveExerciseAction")
+@ActionRegistration(displayName = "#CTL_DownloadAdaptiveExerciseAction")
+@ActionReference(path = "Menu/TM&C", position = -200)
+@Messages("CTL_DownloadAdaptiveExerciseAction=DownloadAdaptiveExerciseAction")
+public final class DownloadAdaptiveExerciseAction implements ActionListener {
+
+    private static final Logger logger = Logger.getLogger(DownloadSolutionAction.class.getName());
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Wrong base class for the logger. Should be: `DownloadAdaptiveExerciseAction`

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public DownloadAdaptiveExerciseAction() {
+        this.projectMediator = ProjectMediator.getInstance();
+        this.dialogs = ConvenientDialogDisplayer.getDefault();
+        this.eventBus = TmcEventBus.getDefault();
+    }
+    
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        // TODO implement action body
+        logger.log(Level.WARNING, "hey you pålk,gpodfk,pogdkf,g");
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Weird language.

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        ProgressObserver observer = new BridgingProgressObserver();
+        Callable<Exercise> ex = TmcCore.get().downloadAdaptiveExercise(observer);
+        BgTask.start("Downloading new adaptive exercise...", ex, observer, new BgTaskListener<Exercise>() {
+            @Override
+            public void bgTaskReady(Exercise ex) {
+                if (ex == null) {
+                    dialogs.displayMessage("Ei adaptiivisia tehtäviä saatavilla :)");
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Weird language.

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                    return;
+                }
+                dialogs.displayMessage("Ladattiin uusi adaptiivinen tehtävä :)");
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Weird language.

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                projectMediator.openProject(proj);
+            }
+
+            @Override
+            public void bgTaskCancelled() {
+            }
+
+            @Override
+            public void bgTaskFailed(Throwable ex) {
+                logger.log(Level.SEVERE, "Something went wrong.");
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Maybe also include the exception as a third argument?

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                    dialogs.displayMessage("Ei adaptiivisia tehtäviä saatavilla :)");
+                    return;
+                }
+                dialogs.displayMessage("Ladattiin uusi adaptiivinen tehtävä :)");
+                TmcProjectInfo proj = projectMediator.tryGetProjectForExercise(ex);
+                projectMediator.openProject(proj);
+            }
+
+            @Override
+            public void bgTaskCancelled() {
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Maybe log this?

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+@ActionID(category = "TMC", id = "fi.helsinki.cs.tmc.actions.DownloadAdaptiveExerciseAction")
+@ActionRegistration(displayName = "#CTL_DownloadAdaptiveExerciseAction")
+@ActionReference(path = "Menu/TM&C", position = -200)
+@Messages("CTL_DownloadAdaptiveExerciseAction=DownloadAdaptiveExerciseAction")
+public final class DownloadAdaptiveExerciseAction implements ActionListener {
+
+    private static final Logger logger = Logger.getLogger(DownloadSolutionAction.class.getName());
+    
+    private ProjectMediator projectMediator;
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: These fields should be final.

File: nbproject/platform.properties [hunk 9]
```diff
@@ -36,4 +36,4 @@ extcluster.mavenmodules.sources=
 #    org.openide.execution,\
 #    org.openide.options,\
 #    org.openide.util.enumerations
-nbplatform.active=nb81
+nbplatform.active=nb82
```
Comment (nygrenh) [inline nbproject/platform.properties]: Please don't include this in your commits.

File: tmc-plugin/nbproject/genfiles.properties [hunk 10]
```diff
@@ -3,6 +3,6 @@ build.xml.script.CRC32=84aa6cb9
 build.xml.stylesheet.CRC32=a56c6a5b@1.45.1
 # This file is used by a NetBeans-based IDE to track changes in generated files such as build-impl.xml.
 # Do not edit this file. You may delete it but then the IDE will never regenerate such files for you.
-nbproject/build-impl.xml.data.CRC32=15b85191
+nbproject/build-impl.xml.data.CRC32=83609e03
```
Comment (nygrenh) [inline tmc-plugin/nbproject/genfiles.properties]: Please don't include this.

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package fi.helsinki.cs.tmc.actions;
+
+import com.google.common.base.Function;
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: unused import

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.util.concurrent.Callable;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.openide.awt.ActionID;
+import org.openide.awt.ActionReference;
+import org.openide.awt.ActionRegistration;
+import org.openide.util.NbBundle.Messages;
+import org.netbeans.api.project.ui.OpenProjects;
```
Comment (nygrenh) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: unused import

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java (hunk 7 above)
Comment (fogre) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadAdaptiveExerciseAction.java]: Logged.

File: tmc-plugin/nbproject/genfiles.properties (hunk 10 above)
Comment (fogre) [inline tmc-plugin/nbproject/genfiles.properties]: Don't include file or changes? This is automatically generated by netbeans.

Comment (nygrenh) [inline tmc-plugin/nbproject/genfiles.properties]: Changes. I recommend reverting this change on your computer.

File: nbproject/platform.properties [hunk 13]
```diff
@@ -1,15 +1,12 @@
...
-    ${nbplatform.active.dir}/cnd:\
     ${nbplatform.active.dir}/extide:\
-    ${nbplatform.active.dir}/dlight:\
     ${nbplatform.active.dir}/harness:\
     ${nbplatform.active.dir}/ide:\
     ${nbplatform.active.dir}/java:\
     ${nbplatform.active.dir}/javafx:\
     ${nbplatform.active.dir}/nb:\
     ${nbplatform.active.dir}/platform:\
-    ${nbplatform.active.dir}/webcommon:\
```
Comment (nygrenh) [inline nbproject/platform.properties]: This is also something Netbeans does sometimes automatically and it should not be in commits.

**Pull request closed without merging**